tensorflow/python/ops/embedding_ops.py

```diff
@@ -935,20 +993,25 @@ def safe_embedding_lookup_sparse(embedding_weights,
       result = array_ops.where(
           is_row_empty, array_ops.zeros_like(result), result, name=scope)
 
-    # Reshape back from linear ids back into higher-dimensional dense result.
-    final_result = array_ops.reshape(
-        result,
-        array_ops.concat([
-            array_ops.slice(
-                math_ops.cast(original_shape, dtypes.int32), [0],
-                [original_rank - 1]),
-            array_ops.slice(array_ops.shape(result), [1], [-1])
-        ], 0))
-    final_result.set_shape(
-        tensor_shape.unknown_shape(
-            (tensor_shape.Dimension(original_rank_dim) - 1).value).concatenate(
-                result.get_shape()[1:]))
-    return final_result
+    if isinstance(sparse_ids, sparse_tensor.SparseTensor):
+      # Reshape back from linear ids back into higher-dimensional dense result.
+      final_result = array_ops.reshape(
+          result,
+          array_ops.concat([
+              array_ops.slice(
+                  math_ops.cast(original_shape, dtypes.int32), [0],
+                  [original_rank - 1]),
+              array_ops.slice(array_ops.shape(result), [1], [-1])
+          ], 0))
+      final_result.set_shape(
+          tensor_shape.unknown_shape(
+              (tensor_shape.Dimension(original_rank_dim) - 1).value)
+                  .concatenate(
+                          result.get_shape()[1:]))
+      return final_result
+    else:
+      # Higher-dimensional case not supported for RaggedTensor input.
+      return result
 
 
 def embedding_lookup_ragged(embedding_weights,
```

Comment:
Sorry for the late reply and thanks for your patience. A couple of high level questions.

1. The if/else conditions in `embedding_lookup_sparse` and `safe_embedding_lookup_sparse` increases code maintenance cost. Does it make sense to move the ragged tensor specific changes to `embedding_lookup_ragged`? For the sharable code in `embedding_lookup_sparse` and `safe_embedding_lookup_sparse`, we could move them out into standalone functions. 
2. Since the main benefit of this CL is performance, could you add some micro-benchmark tests? We have some existing benchmark results at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/eager/benchmarks_test.py.

Reply:
I think the functionality could easily be moved to `embedding_lookup_ragged` or a new function, but some conditionals would probably be needed as users would presumably still call `embedding_lookup_sparse`/`safe_embedding_lookup_sparse`?

I added two simple benchmarks that compare the performance of `embedding_lookup_sparse` with SparseTensor and RaggedTensor inputs to benchmarks_test.py. On an A100 system, the RaggedTensor case runs approx. 8 times faster.

Reply:
Thanks! That's a lot of work!

Some thoughts about the API.

Currently if users call `embedding_lookup` with a RaggedTensor ids, it is redirected to `embedding_lookup_ragged`. If the input is a RaggedTensor, I would expect users to call `embedding_lookup` instead of `embedding_lookup_sparse` or `safe_embedding_lookup_sparse`. If we add RaggedTensor support to `embedding_lookup_sparse`, I imagine users could get confused about whether to use  `embedding_lookup` or `embedding_lookup_sparse` for RaggedTensor. So I would prefer improving `embedding_lookup_ragged` instead of adding RaggedTensor support to `embedding_lookup_sparse`.

TF has a few APIs ends with "_sparse" that are specific for SparseTensors. But for RaggedTensors and other user customized ExtensionTypes, we tend to reuse the standard TensorFlow APIs. For this purpose, we developed  [dispatch_for_api](https://www.tensorflow.org/api_docs/python/tf/experimental/dispatch_for_api). (Note: `embedding_lookup` is not using this API right now.)

I kinda like the dispatch method and I think if possible we should also move SparseTensor to the dispatch method so we could call `embedding_lookup` with `Tensor`, `RaggedTensor` or `SparseTensor`. (Though I don't a specific timeline to make that happen.).

Let me know what you think.

Reply:
There are some functionality differences between `embedding_lookup` and `embedding_lookup_sparse` that would complicate this to some extent. Currently, `embedding_lookup` behaves the same when called with Tensor and RaggedTensor inputs. However, `embedding_lookup_sparse` applies a reduction specified by the `combiner` keyword argument and provides support for weights. Both features are not supported by current `embedding_lookup` and would become available only when the input is a RaggedTensor with a rank of 2. The additional functionality of `safe_embedding_lookup_sparse`, which would presumably be activated by an added flag, would as well only be available when the lookup is called with a compatible RaggedTensor. As such, the functionality of `embedding_lookup` would vary depending on the type of input.

Reply:
Yeah, agreed, these are very good thinkings. 

My feeling is `embedding_lookup` should only do the embedding lookup part of job. The `weights` and `combiner` logic is pretty much independent from the `embedding_lookup`. They could either be handled by user customized function or a standalone TF API if necessary.

In terms of the `allow_dense_grads` flag, my personal opinion is it could make sense to be added as additional flag to `embedding_lookup`, though any changes to API will require API leads approval.

Reply:
The combiner isn't really independent of the embedding lookup. `embedding_lookup_sparse` implements the actual [lookup](https://github.com/tensorflow/tensorflow/blob/45a2cb0250c1517c62a32b99473b9a36acb0ea2a/tensorflow/python/ops/embedding_ops.py#L599) via `math_ops.sparse_segment_sum` instead of calling gather. Changing this implementation would likely affect performance negatively.

Because of the difference in functionality of `embedding_lookup` and `embedding_lookup_sparse`/`safe_embedding_lookup_sparse`, I don't think there is a meaningful risk that enabling RaggedTensor inputs for both types of lookups would confuse users. It seems that the possibility of using RaggedTensor inputs isn't even clearly stated in the [documentation](https://www.tensorflow.org/api_docs/python/tf/nn/embedding_lookup) of `embedding_lookup`.

Reply:
Thanks for the explanation and raising the concern of RaggedTensor documentation. We have a plan in next year to improve RaggedTensor documentation. Here are my thoughts about `embedding_lookup_sparse`.

Ideally the API should be atomic to keep the API surface small. APIs like `embedding_lookup_sparse` do multiple things are generally discouraged. Fusion of multiple small ops should be handled by graph transformation or compiler, not by creating individual composite APIs. So personally I lean to not increasing the maintenance load and code base complexity by _directly_ adding `RaggedTensor` implementation to `embedding_lookup_sparse`.

But I do acknowledge `embedding_lookup_sparse` and `safe_embedding_lookup_sparse` are widely used and `RaggedTensor` support of these APIs brings potential performance benefit. One solution here is to use the type based dispatch API [`tf.experimental.dispatch_for_api`](https://www.tensorflow.org/api_docs/python/tf/experimental/dispatch_for_api) to add a `RaggedTensor` implementation for `embedding_lookup_sparse`. This is a (experimental) public API. We have already used it for most `RaggedTensor` [APIs](https://github.com/tensorflow/tensorflow/tree/e383266c782e4bfaf6c18ffb91b94eb5f44155d1/tensorflow/python/ops/ragged) where you can find a lot of examples. The benefit is that we don't need to modify TF code base. For the `fill_empty_rows custom` op, you could follow guidelines for [custom ops](https://www.tensorflow.org/guide/create_op). I think most work in this CL could be re-used. 

Let me know if you have any questions.